src/expint.jl

```diff
@@ -0,0 +1,410 @@
+import Polynomials
+using Base.MathConstants
+
+#function polymult(p1::Vector{T}, p2::Vector{T}) where {T}
+#    n, m = length(p1) - 1, length(p2) - 1
+#    c = zeros(T, m + n + 1)
+#    for i in 0:n, j in 0:m
+#        c[i + j + 1] += p1[i + 1] * p2[j + 1]
+#    end
+#    return c
+#end
+
+function E₁_cfpoly_approx(n::Integer, pstart::Polynomials.Polynomial{T}, ::Type{T}=BigInt) where {T<:Real}
+    q = Polynomials.Polynomial(T[1])
+    p = pstart
+    x = Polynomials.Polynomial(T[0,1])
+    for i = n:-1:1
+        p, q = x*p+(1+i)*q, p # from cf = x + (1+i)/cf = x + (1+i)*q/p
+        p, q = p + i*q, p     # from cf = 1 + i/cf = 1 + i*q/p
+    end
+    # do final 1/(x + inv(cf)) = 1/(x + q/p) = p/(x*p + q)
+    return p, x*p + q
+end
+
+macro E₁_cf64(x, n::Integer, start)
+    pstart = Polynomials.Polynomial(eval(start))
+    # consider using BigFloat?
+    p, q = E₁_cfpoly_approx(n, pstart, Float64)
+    xesc = esc(x)
+    
+    num_expr =  :(@evalpoly $xesc)
+    append!(num_expr.args, Float64.(Polynomials.coeffs(p)))
+    den_expr = :(@evalpoly $xesc)
+    append!(den_expr.args, Float64.(Polynomials.coeffs(q)))
+    :( exp(-$xesc) * $num_expr / $den_expr )
+end
+
+
+function E₁_taylor_coefficients(::Type{T}, n::Integer) where {T<:Number}
+    n < 0 && throw(ArgumentError("$n ≥ 0 is required"))
+    n == 0 && return T[]
+    n == 1 && return T[-eulergamma]
+    # iteratively compute the terms in the series, starting with k=1
+    term::T = 1
+    terms = T[-eulergamma, term]
+    for k=2:n
+        term = -term * (k-1) / (k * k)
+        push!(terms, term)
+    end
+    return terms
+end
+
+# inline the Taylor expansion for a given order n, in double precision
+macro E₁_taylor64(z, n::Integer)
+    c = E₁_taylor_coefficients(Float64, n)
+    zesc = esc(z)
+    taylor = :(@evalpoly $zesc)
+    append!(taylor.args, c)
+    :( $taylor - log($zesc) )
+end
+
+# adapted from Steven G Johnson's initial implementation: issue #19
+function E₁(x::Float64)
+    x < 0 && throw(DomainError(x, "negative argument, convert to complex first"))
+    x == 0 && return Inf
+    if x > 2.15
+        # specially chosen approximants for faster convergence
+        x < 3.0   && return @E₁_cf64(x, 18, [6.267445506556548, -2.278962735947262, 0.5577231261815463, -0.05413049191473329])
+        x < 4.0   && return @E₁_cf64(x, 16, [5.114292670961982, -1.2789140459431323, 0.22066200334871455, -0.015067049382830766])
+        x < 6.1   && return @E₁_cf64(x, 14, [4.194988480897909, -0.7263593325667503, 0.08956574399359891, -0.00434973529065973])
+        x < 8.15  && return @E₁_cf64(x, 9,  [3.0362016309948228, -0.33793806630590445, 0.029410409377178114, -0.0010060498260648586])
+        x < 25.0  && return @E₁_cf64(x, 8,  [2.5382065303376895, -0.18352177433259526, 0.011141562002742184, -0.0002634921890930066])
```

Comment:
Maybe just use ratfn_minimax from https://github.com/simonbyrne/Remez.jl/blob/master/src/Remez.jl?

Reply:
ApproxFun claims that a degree-16 polynomial gives machine precision over the 2.15..3.0 interval, for example, and presumably a minimax function could do even better.

Reply:
Note that you will want to pass a weighting function of 1/E₁ to `ratfn_minimax` so that it minimizes the maximum *relative* error.